hail/python/hail/backend/spark_backend.py

```diff
@@ -135,14 +135,20 @@ def __init__(self, idempotent, sc, spark_conf, app_name, master,
 
             jars = [hail_jar_path]
 
-            if os.environ.get('HAIL_SPARK_MONITOR'):
+            if os.environ.get('HAIL_SPARK_MONITOR') or os.environ.get('AZURE_SPARK') == '1':
                 import sparkmonitor
                 jars.append(os.path.join(os.path.dirname(sparkmonitor.__file__), 'listener.jar'))
                 conf.set("spark.extraListeners", "sparkmonitor.listener.JupyterSparkMonitorListener")
 
             conf.set('spark.jars', ','.join(jars))
-            conf.set('spark.driver.extraClassPath', ','.join(jars))
-            conf.set('spark.executor.extraClassPath', './hail-all-spark.jar')
+            if os.environ.get('AZURE_SPARK') == '1':
+                print('AZURE_SPARK environment variable is set to "1", assuming you are in HDInsight.')
+                # Setting extraClassPath in HDInsight overrides the classpath entirely so you can't
+                # load the Scala standard library. Interestingly, setting extraClassPath is not
```

Comment:
huh, weird. How do we pick up the sparkmonitor jar then?

Reply:
So, this reminds me of a frustration I had years ago. The Spark docs stated that adding a JAR to the `spark.jars` configuration parameter *should* be sufficient to get the JAR onto the class path. In practice, this was not true. In particular, we added this bizarre `spark.executor.extraClassPath=./hail-all-spark.jar` setting because we knew the JAR file would be copied into the worker directory of the executor.

HDInsight seems to be broken in a slightly different way: the "extra" class path is actually the "whole" class path. Luckily the `spark.jars` parameter works as advertised in HDInsight (it both copies the JAR and modifies the class path).

Reply:
Erm, huh, I realize now I answered a different question. I have no idea how we find the spark monitor JAR.